src/incogniaApi.ts

```diff
@@ -26,226 +17,189 @@ import {
   WebSignupResponse,
   WebTransactionResponse
 } from './types'
-import { buildUserAgent } from './utils'
+import { TokenManager } from './token'
+import { apiEndpoints } from './endpoints'
+import { requestResource } from './request'
 
 type IncogniaApiConstructor = {
   clientId: string
   clientSecret: string
 }
 
-type IncogniaToken = {
-  createdAt: number
-  expiresIn: number
-  accessToken: string
-  tokenType: string
+const errorMessages = {
+  CLIENT_ID: 'Missing required parameter: clientId',
+  CLIENT_SECRET: 'Missing required parameter: clientSecret',
+  INSTALLATION_ID: 'Missing required parameter: installationId',
+  SESSION_TOKEN: 'Missing required parameter: sessionToken',
+  ACCOUNT_ID: 'Missing required parameter: accountId',
+  EVENT: 'Missing required parameter: event',
+  INIT: 'IncogniaApi not initialized'
 }
 
-type ApiEndpoints = {
-  TOKEN: string
-  SIGNUPS: string
-  TRANSACTIONS: string
-  FEEDBACKS: string
-}
+export class IncogniaApi {
+  /*
+   ** Singleton
+   */
+  private static instance?: IncogniaApi
 
-const BASE_ENDPOINT = 'https://api.incognia.com/api'
+  /*
+   ** Instance properties
+   */
+  readonly tokenManager: TokenManager
 
-export const apiEndpoints: ApiEndpoints = {
-  TOKEN: `${BASE_ENDPOINT}/v2/token`,
-  SIGNUPS: `${BASE_ENDPOINT}/v2/onboarding/signups`,
-  TRANSACTIONS: `${BASE_ENDPOINT}/v2/authentication/transactions`,
-  FEEDBACKS: `${BASE_ENDPOINT}/v2/feedbacks`
-}
+  /*
+   ** Initialization
+   */
+  private constructor({ clientId, clientSecret }: IncogniaApiConstructor) {
+    this.tokenManager = new TokenManager({ clientId, clientSecret })
+  }
 
-export class IncogniaApi {
-  readonly clientId: string
-  readonly clientSecret: string
-  incogniaToken: IncogniaToken | null
+  public static init({ clientId, clientSecret }: IncogniaApiConstructor): void {
+    if (!IncogniaApi.instance) {
+      if (!clientId) throw new IncogniaError(errorMessages.CLIENT_ID)
+      if (!clientSecret) throw new IncogniaError(errorMessages.CLIENT_SECRET)
 
-  constructor({ clientId, clientSecret }: IncogniaApiConstructor) {
-    if (!clientId || !clientSecret) {
-      throw new IncogniaError('No clientId or clientSecret provided')
+      IncogniaApi.instance = new IncogniaApi({ clientId, clientSecret })
     }
-
-    this.clientId = clientId
-    this.clientSecret = clientSecret
-    this.incogniaToken = null
   }
 
   /*
-   ** Resources
+   ** Static Methods
    */
-  async registerSignup(props: RegisterSignupProps): Promise<SignupResponse> {
+  public static async registerSignup(
+    props: RegisterSignupProps
+  ): Promise<SignupResponse> {
     const { installationId } = props || {}
-    if (!installationId) {
-      throw new IncogniaError('No installationId provided')
-    }
+    if (!IncogniaApi.instance) throw new IncogniaError(errorMessages.INIT)
```

Comment:
I was thinking: it would be nice to have a decorator function that performs this check. I think we could avoid a lot of repetition. But I realize there are not a lot of functions and the check is simple. Do you think it would be worth it?

Reply:
Pensei em não fazer só pelo código de check ser simples mesmo. Acho que seria parecido em termos de complexidade.